src/make.ts

```diff
@@ -95,15 +95,58 @@ function error<T>(error: string): Make<T> {
   return Make.error<T>([{ path: [], error }]);
 }
 
-function checkString(value: any) {
-  if (typeof value !== 'string') {
-    return getErrorWithValueMsg('expected a string', value);
+export function registerFormat(format: string, maker: Maker<any, undefined>) {
+  assert(!formats[format], `format ${format} is already registered`);
+  formats[format] = maker;
+}
+
+const formats: Record<string, Maker<any, undefined>> = {};
+
+function getFormatter(format: string | undefined): Maker<any, undefined> {
+  if (format === undefined) {
+    return () => Make.ok(undefined);
   }
-  return Make.ok(value);
+  const formatter = formats[format];
+  if (!formatter) {
+    return () => Make.ok(undefined);
+  }
+  return formatter;
+}
+
+function getPatterner(pattern: string | undefined): Maker<any, void> {
+  if (pattern === undefined) {
+    return value => Make.ok(value);
+  }
+  const reg = new RegExp(pattern);
+  return value => {
+    if (!reg.test(value)) {
+      return error(`${value} does not match pattern /${pattern}/`);
+    }
+    return Make.ok(undefined);
+  };
 }
 
-export function makeString() {
-  return checkString;
+export function makeString(
+  format: string | undefined = undefined,
+  pattern: string | undefined = undefined
+): Maker<any, string> {
+  const formatter = getFormatter(format);
+  const patterner = getPatterner(pattern);
+
+  return (value: any) => {
+    if (typeof value !== 'string') {
+      return getErrorWithValueMsg('expected a string', value);
+    }
+    const formatted = formatter(value);
+    if (formatted.isError()) {
+      return formatted;
+    }
+    const patterned = patterner(value);
```

Comment:
patterned, patterner
![iu-1](https://user-images.githubusercontent.com/5294529/79334547-2a66fa80-7f29-11ea-8f92-3e19da973482.png)

Reply:
do I detect a hint of well placed awe on my naming choices?